File: app/src/main/java/mydebts/android/app/data/internal/ClosableCursorToList.java
package mydebts.android.app.data.internal;

import android.database.Cursor;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.functions.Function;

class ClosableCursorToList<T> implements Function<Cursor, List<T>> {
    private final Function<Cursor, T> itemMapper;

    ClosableCursorToList(Function<Cursor, T> itemMapper) {
        this.itemMapper = itemMapper;
    }

    @Override
    public List<T> apply(Cursor cursor) throws Exception {
        List<T> ts = new ArrayList<>();

        for (cursor.moveToFirst(); !cursor.isAfterLast(); cursor.moveToNext()) {
            ts.add(itemMapper.apply(cursor));
        }

        cursor.close();

        return ts;
    }
}

File: app/src/main/java/mydebts/android/app/data/internal/ClosableCursorToList.kt
package mydebts.android.app.data.internal

import android.database.Cursor

import java.util.ArrayList

import io.reactivex.functions.Function

internal class ClosableCursorToList<T>(private val itemMapper: Function<Cursor, T>) : Function<Cursor, List<T>> {

    @Throws(Exception::class)
    override fun apply(cursor: Cursor): List<T> {
        val ts = ArrayList<T>()

        cursor.moveToFirst()
        while (!cursor.isAfterLast) {
            ts.add(itemMapper.apply(cursor))
            cursor.moveToNext()
        }

        cursor.close()

        return ts
    }
}
